static/js/message_fetch.js

```diff
@@ -500,6 +501,11 @@ export function initialize(home_view_loaded) {
         filter: new Filter([{operator: "in", operand: "home"}]),
         excludes_muted_topics: true,
     });
+    // TODO: Ideally we'd have loading indicators for recent topics at
+    // both top and bottom be managed by load_messages, but that
+    // likely depends on other reorganizations of the early loading
+    // sequence.
+    loading.make_indicator($("#recent_topics_loading_messages_indicator"));
```

Comment:
OK new proposal; we can actually fix the messyness in `recent_topics_ui.js` if we just pass a `cont` parameter (continuation) to hide the loading indicator in the `load_messages` call on the next line.

I think that lets us no longer need the `initial_loading_indicator_hidden` parameter and various other elements of complexity, and would also avoid the "private messages" issue you mentioned, since structurally we're just hiding the loading indicator once we've fetched the initial data set for this feature.